zcash_primitives/src/transaction/components/sapling/builder.rs

```diff
@@ -373,204 +462,401 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
+}
+
+/// The parts needed to sign a [`SpendDescription`].
+#[derive(Clone, Debug)]
+pub struct SigningParts {
+    /// The spend validating key for this action. Used to match spend authorizing keys to
+    /// actions they can create signatures for.
+    ak: PublicKey,
+    /// The randomization needed to derive the actual signing key for this note.
+    alpha: jubjub::Scalar,
+}
+
+/// Marker for a partially-authorized bundle, in the process of being signed.
+#[derive(Debug)]
+pub struct PartiallyAuthorized {
+    binding_signature: Signature,
+    sighash: [u8; 32],
+}
+
+impl InProgressSignatures for PartiallyAuthorized {
+    type AuthSig = MaybeSigned;
+}
+
+/// A heisen[`Signature`] for a particular [`SpendDescription`].
+#[derive(Clone, Debug)]
+pub enum MaybeSigned {
+    /// The information needed to sign this [`SpendDescription`].
+    SigningMetadata(SigningParts),
+    /// The signature for this [`Action`].
+    Signature(Signature),
+}
+
+impl MaybeSigned {
+    fn finalize(self) -> Result<Signature, Error> {
+        match self {
+            Self::Signature(sig) => Ok(sig),
+            _ => Err(Error::MissingSignatures),
+        }
+    }
+}
+
+impl<P: InProgressProofs> Bundle<InProgress<P, Unsigned>> {
+    /// Loads the sighash into this bundle, preparing it for signing.
+    ///
+    /// This API ensures that all signatures are created over the same sighash.
+    pub fn prepare<R: RngCore + CryptoRng>(
+        self,
+        mut rng: R,
+        sighash: [u8; 32],
+    ) -> Bundle<InProgress<P, PartiallyAuthorized>> {
+        self.map_authorization((
+            |proof| proof,
+            |proof| proof,
+            |parts| MaybeSigned::SigningMetadata(parts),
+            |auth: InProgress<P, Unsigned>| InProgress {
+                sigs: PartiallyAuthorized {
+                    binding_signature: auth.sigs.bsk.sign(
+                        &sighash,
+                        &mut rng,
+                        VALUE_COMMITMENT_RANDOMNESS_GENERATOR,
+                    ),
+                    sighash,
+                },
+                _proof_state: PhantomData::default(),
+            },
+        ))
+    }
+}
+
+impl Bundle<InProgress<Proven, Unsigned>> {
+    /// Applies signatures to this bundle, in order to authorize it.
+    ///
+    /// This is a helper method that wraps [`Bundle::prepare`], [`Bundle::sign`], and
+    /// [`Bundle::finalize`].
+    pub fn apply_signatures<R: RngCore + CryptoRng>(
+        self,
+        mut rng: R,
+        sighash: [u8; 32],
+        signing_keys: &[PrivateKey],
+    ) -> Result<Bundle<Authorized>, Error> {
+        signing_keys
+            .iter()
+            .fold(self.prepare(&mut rng, sighash), |partial, ask| {
+                partial.sign(&mut rng, ask)
+            })
+            .finalize()
+    }
+}
+
+impl<P: InProgressProofs> Bundle<InProgress<P, PartiallyAuthorized>> {
+    /// Signs this bundle with the given [`SpendAuthorizingKey`].
```

Comment:
The comment here says `SpendAuthorizingKey` but the type of the argument is `PrivateKey`; this is confusing.

Reply:
Copy-pasta from `orchard`.